### DISCUSSION TITLE
Performance / tracing

### DISCUSSION THREAD

**Discussion by mikeaustin** [peggyjs/peggy#428] [General]
First, I want to say I'm glad peggy is still chugging a long and there is new activity and new versions. I love it and the former pegjs for it's simplicity and for it's tight JS integration which requires much less boilerplate.

I've re-written my language (Kopi) many times in Peg**, and sometimes I hit a performance issue. In my current version, there is a 16 × difference in speed in parsing with * or ? vs using a "/". I can get around it, but the rules are not as clean.

I believe what is happening is that using the "*" and "?", the PrimaryExpression can be re-used, while using "/" the expression is discarded and the alternate expression needs to be parsed again. Is that a correct assumption?

```
RangeExpression
  = from:PrimaryExpression _ ".." _ to:PrimaryExpression {
      return  !expr ? from : {
        type: 'RangeExpression',
        from,
        to: expr[3],
      };
    }
  / PrimaryExpression
```
vs
```
RangeExpression2
  = from:PrimaryExpression expr:(_ ".." _ PrimaryExpression)? {
      return  !expr ? from : {
        type: 'RangeExpression',
        from,
        to: expr[3],
      };
    }
```
Should I just always use the "*" and "?" style for precedence rules?

**Comment by hildjj** [comment]
You are correct that the version with `/` has to re-parse PrimaryExpression if the first alternative doesn't match.  Peggy always discards and goes back to the point where parsing started for the previous alternative if that alternative doesn't match.  

How complicated is PrimaryExpression?

**Reply by mikeaustin** [reply]
Each rule is not very complicated, but there are many chained for precedence like most languages. With one test, it never finished parsing, or at least it was taking a long time. I’ve converted to the * / where I can and it is much faster. I remember this same situation a few years ago, I just couldn’t remember the specifics.

This is a temporary home for the new version of Kopi, my for fun/research language. You can try it live in a REPL at https://mike-austin.com/react-desktop2

https://github.com/GEMTASK/shared-components/blob/main/test/src/components/terminal/kopi-language/src/parser.pegjs
